streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -222,60 +230,109 @@ public void flush() {
             }
             dirtyKeys.clear();
         }
     }
 
+    private void logValue(final Bytes key, final BufferKey bufferKey, final ContextualRecord value) {
+        final byte[] serializedContextualRecord = value.serialize();
+
+        final byte[] timeAndContextualRecord = ByteBuffer.wrap(new byte[Long.BYTES + serializedContextualRecord.length])
+                                                         .putLong(bufferKey.time)
+                                                         .put(serializedContextualRecord)
+                                                         .array();
+
+        collector.send(
+            changelogTopic,
+            key,
+            timeAndContextualRecord,
+            V_1_CHANGELOG_HEADERS,
+            partition,
+            null,
+            KEY_SERIALIZER,
+            VALUE_SERIALIZER
+        );
+    }
+
+    private void logTombstone(final Bytes key) {
+        collector.send(changelogTopic,
+                       key,
+                       null,
+                       null,
+                       partition,
+                       null,
+                       KEY_SERIALIZER,
+                       VALUE_SERIALIZER
+        );
+    }
+
     private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch) {
         for (final ConsumerRecord<byte[], byte[]> record : batch) {
             final Bytes key = Bytes.wrap(record.key());
             if (record.value() == null) {
                 // This was a tombstone. Delete the record.
                 final BufferKey bufferKey = index.remove(key);
                 if (bufferKey != null) {
-                    sortedMap.remove(bufferKey);
+                    final ContextualRecord removed = sortedMap.remove(bufferKey);
+                    if (removed != null) {
+                        memBufferSize -= computeRecordSize(bufferKey.key, removed);
+                    }
+                    if (bufferKey.time == minTimestamp) {
+                        final BufferKey leastKey = sortedMap.firstKey();
+                        minTimestamp = leastKey != null ? leastKey.time : Long.MAX_VALUE;
+                    }
+                }
+
+                if (record.partition() != partition) {
+                    throw new IllegalStateException(
+                        String.format(
+                            "record partition [%d] is being restored by the wrong suppress partition [%d]",
+                            record.partition(),
+                            partition
+                        )
+                    );
                 }
             } else {
                 final ByteBuffer timeAndValue = ByteBuffer.wrap(record.value());
                 final long time = timeAndValue.getLong();
                 final byte[] value = new byte[record.value().length - 8];
                 timeAndValue.get(value);
+                if (record.headers() == null || record.headers().lastHeader("v") == null) {
```

Comment:
If we are serializing the `ContextualRecord` could we ever hit this case?

Reply:
Yep, the currently released versions of Suppress are already serializing changelog messages without this header. The purpose of this check is to be able to transition to new Streams versions in a backward-compatible way.

Reply:
If it's an "old format" changelog record, the headers would have been copied from the record context, in which case they could be null or non-null, but still not contain a "v". Of course, it could contain a "v" which has nothing to do with our algorithm here, but I think that situation is undetectable. And unlikely, as well.

Reply:
I don't think so. We never write headers in the changelogger. Note, that the changelog topic is used to recover the store content. However, rows in a store only have a key and a value. There is no header that we could write, because the on put, the current record header does not related to the store content.

Similarly, `suppress()` serializes the whole record context and store it in the value IIRC.

Reply:
good call. So, we're covered. Thanks for this analysis.

Reply:
Well, my point is, that the check can be simplified. I don't think that `record.headers() == null` can be true; it's guaranteed that there is a headers object.

Not sure if we can simplify the second check. It iterators over the headers map and does String comparison to find a header with key `v` -- seems to be rather heavy.

Reply:
If I understand correct, a record read from the changelog topic should only be:

1) having no headers at all (old version)
2) having a singleton header with `v --> byte`.

All other cases should indicate a bug in the code.

So it seems we can just check if `record.headers() == null`, and inside the if condition though, we should also check the `v` and assume it's always there (if not then throw illegal state), and switching on the byte value:

1) byte == 1: do as below.
2) otherwise: to not support forward compatibility, we can just throw unsupported.

Reply:
We were previously setting the changelog record headers from the record headers. I really don't think it's safe to just assume that record headers are always empty for records inside of Streams.

Also, I'd like to point out that @guozhangwang thinks that old-format headers are _always_ null, and @mjsax think that they are _never_ null, but _always_ empty...

The point is, header handling is partially undefined inside Streams. I'd be very much in favor of being defensive in the face of such ambiguity, rather than making potentially faulty assumptions, since the risk is a runtime exception or even data corruption (if we mis-interpret the binary format, but it happens to de-serialize without exceptions into gibberish).

Further, none of this feedback is pointing out a problem with the code, just that there might be one unnecessary clause in the boolean expression (which, I still am not convinced is unnecessary after all).

Reply:
Thanks for the explanation @vvcephei , makes sense.

One thing I'd still like to ask though: from the current code it seems we do not check the byte of `v` if it indeed exist, while I think to be safer we should still check it, and throw exception indicating that we do not support forward compatibility when the format has evolved. WDYT?

Reply:
- It's a contract that `KafkaConsumers` _guarantees_ that `header != null`. \cc @hachikuji to confirm.
- And we know that KafkaStreams never writes headers into changelog topics.

Thus, I don't see any reason to check for something that we know is the case, ie, we know that `header.size() == 0` in old format. For everything else, we could throw `IllegalStateException`. Of course the header API is limiting and we cannot get `size()` and thus `record.headers().lastHeader("v") == null)` is what we need to do... :( -- but we can safely remove the first `null` check -- it could even mask a bug and we should rather fail for this case.

We can also do a version check as suggested by Guozhang.

Reply:
> we know that KafkaStreams never writes headers into changelog topics.

We're restoring records that we wrote like this: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java#L240

We write headers to the suppression changelog topic. Sorry if that wasn't clear, that's why I'm insisting we have to check it.

If there's a contract that headers are never null, then I agree we can remove that check. In that case, we should add null-checks to the ConsumerRecord constructor.

Reply:
Ok, I've had a chat with @hachikuji , and he confirmed that the headers _should_ never be null. I did a quick audit of code base as well, and agree that this should be the case.

I added a null-check to the header and docs to this effect in f5bd0999c (in this PR), and then removed the null-check.

I also added an exception if the record has a "v" header, but it isn't `1`. I'm wondering, @guozhangwang , if we should just assume in this case that the record is old-format, and happened to already have a "v" header. The alternative thing is you could have run Streams with a _newer_ version, and then downgraded. In that case, it would be wrong to try restoring as old-format, and an exception would be preferable. I'm leaning toward the exception, since people haven't been using Suppress that long, so the risk of a spurious "v" header is low. WDYT?

Reply:
> We're restoring records that we wrote like this: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java#L240
>
> We write headers to the suppression changelog topic. Sorry if that wasn't clear, that's why I'm insisting we have to check it.

Arg. We should not have done this... But serialize the header in KS and add them to the value. Well, too late now I guess. Or could we switch the format? Note sure what the impact on the upgrade path would be? Thought?

> I added a null-check to the header and docs to this effect in f5bd099 (in this PR), and then removed the null-check.

+100